packages/transformers/js/core/src/runner.rs

```diff
@@ -0,0 +1,187 @@
+use std::string::FromUtf8Error;
+use swc_core::common::input::StringInput;
+use swc_core::common::sync::Lrc;
+use swc_core::common::util::take::Take;
+use swc_core::common::{FileName, Globals, Mark, SourceMap, GLOBALS};
+use swc_core::ecma::ast::Module;
+use swc_core::ecma::codegen::text_writer::JsWriter;
+use swc_core::ecma::parser::lexer::Lexer;
+use swc_core::ecma::parser::Parser;
+use swc_core::ecma::transforms::base::resolver;
+use swc_core::ecma::visit::{Fold, FoldWith, VisitMut, VisitMutWith};
+
+pub struct RunContext {
+  /// Source-map in use
+  pub source_map: Lrc<SourceMap>,
+  /// Global mark from SWC resolver
+  pub global_mark: Mark,
+  /// Unresolved mark from SWC resolver
+  pub unresolved_mark: Mark,
+}
+
+pub struct RunVisitResult<V> {
+  pub output_code: String,
+  #[allow(unused)]
+  pub visitor: V,
+  pub source_map: Vec<u8>,
+}
+
+/// Runner of SWC transformations
+///
+/// * Parse `code` with SWC
+/// * Run a visitor over it
+/// * Return the result
+///
+pub fn run_visit<V: VisitMut>(
+  code: &str,
+  make_visit: impl FnOnce(RunContext) -> V,
+) -> Result<RunVisitResult<V>, RunWithTransformationError> {
+  let (output_code, visitor, source_map) =
+    run_with_transformation(code, |run_test_context: RunContext, module: &mut Module| {
+      let mut visit = make_visit(run_test_context);
+      module.visit_mut_with(&mut visit);
+      visit
+    })?;
+  Ok(RunVisitResult {
+    output_code,
+    visitor,
+    source_map,
+  })
+}
+
+/// Same as `run_visit` but for `Fold` instances
+#[allow(unused)]
+pub fn run_fold<V: Fold>(
+  code: &str,
+  make_fold: impl FnOnce(RunContext) -> V,
+) -> Result<RunVisitResult<V>, RunWithTransformationError> {
+  let (output_code, visitor, source_map) =
+    run_with_transformation(code, |run_test_context: RunContext, module: &mut Module| {
+      let mut visit = make_fold(run_test_context);
+      *module = module.take().fold_with(&mut visit);
+      visit
+    })?;
+  Ok(RunVisitResult {
+    output_code,
+    visitor,
+    source_map,
+  })
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum RunWithTransformationError {
+  #[error("Failed to parse")]
+  SwcParse(swc_ecma_parser::error::Error),
+  #[error("IO Error: {0}")]
+  IoError(#[from] std::io::Error),
+  #[error("Non-utf 8 output: {0}")]
```

Comment:
This can be structured a bit better, depending on the exact context.

```suggestion
  #[error("Invalid utf-8 output: {0}")]
  InvalidUtf8Output(#[from] FromUtf8Error),
```